contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -631,11 +624,65 @@ bool USDScene::hasAttribute( const SceneInterface::Name &name ) const
 void USDScene::attributeNames( SceneInterface::NameList &attrs ) const
 {
 	attrs.clear();
-	auto visibilityAttr = pxr::UsdGeomImageable( m_location->prim ).GetVisibilityAttr();
-	if( visibilityAttr.HasAuthoredValue() )
+	if( m_location->prim.IsPseudoRoot() )
+	{
+		// No attributes here
+		return;
+	}
+
+	pxr::UsdGeomImageable imageable( m_location->prim );
+	if( imageable.GetVisibilityAttr().HasAuthoredValue() )
 	{
 		attrs.push_back( SceneInterface::visibilityName );
 	}
+	if( imageable && imageable.GetPurposeAttr().HasAuthoredValue() )
```

Comment:
Do we need to do this `imageable &&` check for visibility as well or is this not-actually-imageable-but-has-attrs specific to purpose? A naive read of the code looks like imageable might be null, in which case the visibility check above could have thrown.

Reply:
I wondered this and I'm afraid I put it to the back of my mind in a just-do-what-you-came-here-to-do-and-get-out-again sort of a way. But looking through the USD visibility source, it looks to be similar to the purpose code - `UsdGeomImageable.GetVisibilityAttr()` will merrily return one an attribute from a non-imageable, but when they come to compute inherited visibility they ignore those. If it's OK with you, I'd like to save that for a follow up PR - this one was really about squeezing just one more thing into 0.59 while I waited for review on another PR. I'd like to think we're talking about corner cases that won't occur in practice. I'm also going to have to refactor attributes considerably when it comes to supporting custom attributes and arnold attributes, at which point I can investigate more fully.

Reply:
sure